### PR TITLE
Xmlenc support

### PR DESCRIPTION
**Pull request by pkarman**
This PR builds on the work in #41 and adds support for XML encryption via the xmlenc gem.

This branch was based on the work in #41 so evaluate/merge that PR first.

Note that this depends on a pending PR for the xmlenc gem: ~~https://github.com/digidentity/xmlenc/pull/23~~

### CODE REVIEW

**Comment by pkarman** [comment]
@jphenow I realize you are no longer active with sportngin -- is someone else taking over maintenance of this gem? I have a couple large PRs pending and would be good to know if it is reasonable to expect that to generate any conversation.

**Comment by jphenow** [comment]
I'll tap some people to see if they're intending to continue support otherwise I can try to see a few of these through over the weekend or early next week and law down some notes about expectations.

**Comment by jphenow** [comment]
@pkarman when you get a moment, could you sync this with master so it cleans up the diff a bit? Then I'll review this one, all being good we can merge and release a 0.3.0

**Comment by pkarman** [comment]
@jphenow 👍 rebased

File: lib/saml_idp/assertion_builder.rb [hunk 1]
```diff
@@ -73,6 +77,81 @@ def fresh
     alias_method :raw, :fresh
     private :fresh
 
+    def encrypt(opts = {})
```
Comment (jphenow) [inline lib/saml_idp/assertion_builder.rb]: Most of the added code in this class seems to be somewhat self-contained. What do you think about moving that into its own encapsulated class?

File: lib/saml_idp/assertion_builder.rb [hunk 2]
```diff
@@ -14,10 +15,12 @@ class AssertionBuilder
     attr_accessor :raw_algorithm
     attr_accessor :authn_context_classref
     attr_accessor :expiry
+    attr_accessor :encryption_opts
+    attr_accessor :encryption_key
```
Comment (jphenow) [inline lib/saml_idp/assertion_builder.rb]: Does this need to be added to the instance state?

File: saml_idp.gemspec [hunk 3]
```diff
@@ -41,6 +41,7 @@ defaults in a Production environment. Post any issues you to github.
   s.add_dependency('builder')
   s.add_dependency('httparty')
   s.add_dependency('nokogiri')
+  s.add_dependency('xmlenc', '>= 0.6.4')
```
Comment (jphenow) [inline saml_idp.gemspec]: One concern is that the encryption stuff isn't necessary in previous versions and I tend to be sensitive to adding new dependencies (while this one seems entirely worth it in the case where users care to have it encrypted). What's your feel on how necessary or common it is in the SAML community to use what you're introducing? If its not ~80% of SAML users using this technique, perhaps we should make this development dependency and test the libraries presence when we try to run this encryption work you've added.

Not sure I'm making sense, if not I can illustrate a bit with an example.

File: lib/saml_idp/assertion_builder.rb (hunk 1 above)
Comment (pkarman) [inline lib/saml_idp/assertion_builder.rb]: e.g. `class EncryptedAssertionBuilder < AssertionBuilder` ?

File: lib/saml_idp/assertion_builder.rb (hunk 2 above)
Comment (pkarman) [inline lib/saml_idp/assertion_builder.rb]: Not strictly, no. Mostly it's for making testing easier, although now I see that I didn't end up using the test where I needed to access the key. I will remove that accessor.

File: saml_idp.gemspec (hunk 3 above)
Comment (pkarman) [inline saml_idp.gemspec]: In our case (US Government), encrypting personally identifiable information is always required. I can't speak for the SAML users everywhere, but anyone dealing with PII (credit cards, e.g.) should probably be considering encryption, on top of whatever encrypted transport (SSL) they might be using.

Comment (pkarman) [inline saml_idp.gemspec]: We can make it an optional dependency, esp if we split the assertion building into an encryption-specific class, and then load `xmlenc` explicitly in the encrypted assertion class. I'm find with that. Mostly a "install it if you intend to use it" philosophy, with some helpful error messaging.

File: lib/saml_idp/assertion_builder.rb (hunk 1 above)
Comment (jphenow) [inline lib/saml_idp/assertion_builder.rb]: Correct me if I'm wrong but the new code in this class doesn't much use the logic that existed previously and the existing code doesn't rely on this new code. So I was thinking a class onto its own that doesn't need to inerhit assertion builder at all, so in encrypt you'd call `EncryptedAssertionBuilder.new(opts).build`. Does that make sense?

File: saml_idp.gemspec (hunk 3 above)
Comment (jphenow) [inline saml_idp.gemspec]: Yep that's what I'm picturing as well. We can add it to the install message I believe in the gemspec and error when people try to encrypt without the library also installed

**Comment by pkarman** [comment]
@jphenow see if those recent commits align more with what you're thinking.

**Comment by jphenow** [comment]
It seems more like it should be of a service to the original assertion builder itself. Conceptually, its an extension but not in the inheritance fashion, more like a plugin:

```ruby
module SamlIdp
    class AssertionBuilder
      #...
      def encrypt
        Encryptor.new(opts[:sign] ? signed : raw, encryption_opts).encrypt
      end
      #...
    end
end

# try requiring, fail hard on missing library
require 'xmlenc'
module SamlIdp
  class EncryptedAssertionBuilder < AssertionBuilder
    def initialize(body, encryption_opts=nil); end

    def encrypt(body, opts = {})
      encryption_template = Nokogiri::XML::Document.parse(build_encryption_template).root
      encrypted_data = Xmlenc::EncryptedData.new(encryption_template)
      # etc. etc.
    end

    def cert
      if encryption_opts[:cert].is_a?(String)
        cert = OpenSSL::X509::Certificate.new(Base64.decode64(encryption_opts[:cert]))
      else
        cert = encryption_opts[:cert]
      end
    end
    private :cert

    # etc. etc.
  end
end
```

Most of that is used to illustrate the point so don't really use it verbatim, but is there something I'm missing about whether that's reasonable for this usecase? if that encrypt code needs to be more tightly encapsulated with the actual assertion builder we can drop that idea.

**Comment by pkarman** [comment]
@jphenow ok I think I understand better now what you were envisioning. Take a look at my latest version.

**Comment by jphenow** [comment]
Yes this is great! Thanks for following along. 

Could you express in the README how this would work for those who would like to use this in their apps? I also just realized I didn't check on that for the SLO addition either, so perhaps include docs (if applicable) for that as well?

We can get this merged and released then.

**Comment by pkarman** [comment]
@jphenow docs added in bf31bff

The logout example is so simple as to approach hand-waving, and only handles the case of SP-initiated logout, so I'm not sure it is worth including. A real-life example is https://github.com/18F/identity-idp/blob/master/app/controllers/concerns/saml_idp_logout_concern.rb which shows how complex it can be.

**Comment by jphenow** [comment]
Yea good point. While I'm inclined to generally only show the what entry points the library gives you so you can make it as complex as it needs to be, perhaps since your Concern you link to is a real-world, open source example we could link to it noting that as a possible reference more grandeur examples? Let me know what ya think, happy with this too, so either way

**Comment by pkarman** [comment]
I'm hesitant to include a link to that app in the docs since it is quite possible the app will get restructured and that URL breaks.

I'm fine with what bf31bff does -- if it's ok with you, I'm happy to release as is.

**Comment by jphenow** [comment]
If the link is the main concern, you could reference a sha rather than the branch to be safe, but also happy to let people complicate that code themselves in this case. Will release as is

**Comment by jphenow** [comment]
Oh shoot! Could you [add a post install message](http://rubygems.rubyforge.org/rubygems-update/Gem/Specification.html#attribute-i-post_install_message) that explains or links to the readme for the encryption addition and call out to install xmlenc?

**Comment by pkarman** [comment]
postinstall message in 6fb3068

**Comment by jphenow** [comment]
Great, thank you sir. Thanks for keeping up with the comments and seeing this through.

**Comment by jphenow** [comment]
`v0.3.0` released. Thanks @pkarman

**Comment by pkarman** [comment]
lovely, thanks for working this through with me.

**Pull request merged**